packages/agent/src/oidc.ts

```diff
@@ -0,0 +1,709 @@
+  aud: string;
+  /** Time at which the JWT was issued. */
+  iat: number;
+  /** Expiration time on or after which the ID Token MUST NOT be accepted
+   * for processing. */
+  exp: number;
+  /** Time when the End-User authentication occurred. */
+  auth_time?: number;
+  /** b64url encoded nonce used to associate a Client session with an ID Token, and to
+   * mitigate replay attacks. */
+  nonce?: string;
+  /** Custom claims. */
+  [key: string]: any;
+};
+
+/** An auth response that is compatible with both Web5 Connect and (hopefully, WIP) OIDC SIOPv2 */
+export type Web5ConnectAuthResponse = {
+  delegateGrants: any[];
+  delegateDid: PortableDid;
+} & SIOPv2AuthResponse;
+
+/** Represents the different OIDC endpoint types.
+ * 1. `pushedAuthorizationRequest`: client sends {@link PushedAuthRequest} receives {@link PushedAuthResponse}
+ * 2. `authorize`: provider gets the {@link Web5ConnectAuthRequest} JWT that was stored by the PAR
+ * 3. `callback`: provider sends {@link Web5ConnectAuthResponse} to this endpoint
+ * 4. `token`: client gets {@link Web5ConnectAuthResponse} from this endpoint
+ */
+type OidcEndpoint =
+  | 'pushedAuthorizationRequest'
+  | 'authorize'
+  | 'callback'
+  | 'token';
+
+/**
+ * Gets the correct OIDC endpoint out of the {@link OidcEndpoint} options provided.
+ * Handles a trailing slash on baseURL
+ *
+ * @param {Object} options the options object
+ * @param {string} options.baseURL for example `http://foo.com/connect/
+ * @param {OidcEndpoint} options.endpoint the OIDC endpoint desired
+ * @param {string} options.authParam this is the unique id which must be provided when getting the `authorize` endpoint
+ * @param {string} options.tokenParam this is the random state as b64url which must be provided with the `token` endpoint
+ */
+function buildOidcUrl({
+  baseURL,
+  endpoint,
+  authParam,
+  tokenParam,
+}: {
+  baseURL: string;
+  endpoint: OidcEndpoint;
+  authParam?: string;
+  tokenParam?: string;
+}) {
+  switch (endpoint) {
+    /** 1. client sends {@link PushedAuthRequest} & client receives {@link PushedAuthResponse} */
+    case 'pushedAuthorizationRequest':
+      return concatenateUrl(baseURL, 'par');
+    /** 2. provider gets {@link Web5ConnectAuthRequest} */
+    case 'authorize':
+      if (!authParam)
+        throw new Error(
+          `authParam must be providied when building a token URL`
+        );
+
+      return authParam
+        ? concatenateUrl(baseURL, `authorize/${authParam}.jwt`)
+        : '';
+    /** 3. provider sends {@link Web5ConnectAuthResponse} */
+    case 'callback':
+      return concatenateUrl(baseURL, `callback`);
+    /**  4. client gets {@link Web5ConnectAuthResponse */
+    case 'token':
+      if (!tokenParam)
+        throw new Error(
+          `tokenParam must be providied when building a token URL`
+        );
+
+      return tokenParam
+        ? concatenateUrl(baseURL, `token/${tokenParam}.jwt`)
+        : '';
```

Comment:
Sanity: since there is a check `if(!tokenParam)` above, there is no need for this conditional return?

Same goes to `authParam` check above.

Reply:
donw